src/transport_ofi.c

```diff
@@ -622,40 +629,60 @@ int bind_enable_ep_resources(shmem_transport_ctx_t *ctx)
     return ret;
 }
 
-
+#ifdef USE_FI_HMEM
 static inline
-int allocate_recv_cntr_mr(void)
+int ofi_mr_reg_external_heap(void)
 {
     int ret = 0;
-    uint64_t flags = 0;
-
-    /* ------------------------------------ */
-    /* POST enable resources for to EP      */
-    /* ------------------------------------ */
-
-    /* since this is AFTER enable and RMA you must create memory regions for
-     * incoming reads/writes and outgoing non-blocking Puts, specifying entire
-     * VA range */
+    uint64_t key = 2;
+
+    const struct iovec iov = {
+                               .iov_base     = shmem_external_heap_base,
+                               .iov_len      = shmem_external_heap_length
+                             };
+    const struct fi_mr_attr mr_attr = {
+                                        .mr_iov         = &iov,
+                                        .iov_count      = 1,
+                                        .access         = FI_REMOTE_READ | FI_REMOTE_WRITE,
+                                        .requested_key  = key,
+                                        .iface          = (shmem_external_heap_device_type == 
+                                                          SHMEMX_EXTERNAL_HEAP_ZE ? FI_HMEM_ZE : FI_HMEM_CUDA),
+                                        .device.ze      = shmem_external_heap_device, /* TODO: Need to change to local */
+                                        .offset         = 0,
+                                        .context        = NULL
+                                      };
+
+    ret = fi_mr_regattr(shmem_transport_ofi_domainfd, &mr_attr, 0, &shmem_transport_ofi_external_heap_mrfd);
+    OFI_CHECK_RETURN_STR(ret, "fi_mr_regattr (heap) failed");
 
 #if ENABLE_TARGET_CNTR
-    {
-        struct fi_cntr_attr cntr_attr = {0};
-
-        /* Create counter for incoming writes */
-        cntr_attr.events   = FI_CNTR_EVENTS_COMP;
-        cntr_attr.wait_obj = FI_WAIT_UNSPEC;
+    ret = fi_mr_bind(shmem_transport_ofi_external_heap_mrfd,
+                     &shmem_transport_ofi_target_cntrfd->fid,
+                     FI_REMOTE_WRITE);
+    OFI_CHECK_RETURN_STR(ret, "target CNTR binding to external heap MR failed");
 
-        ret = fi_cntr_open(shmem_transport_ofi_domainfd, &cntr_attr,
-                           &shmem_transport_ofi_target_cntrfd, NULL);
-        OFI_CHECK_RETURN_STR(ret, "target CNTR open failed");
+    if (shmem_transport_ofi_info.p_info->domain_attr->mr_mode & FI_MR_ENDPOINT) {
+        ret = fi_ep_bind(shmem_transport_ofi_target_ep,
+                         &shmem_transport_ofi_target_cntrfd->fid, FI_REMOTE_WRITE);
+        OFI_CHECK_RETURN_STR(ret, "target CNTR binding to target EP failed");
+        ret = fi_mr_bind(shmem_transport_ofi_external_heap_mrfd,
```

Comment:
Can you explain what the difference between this **mr_bind** call and the one on line 659 is? I understand that the **fid\*** argument is different for each. Is it that, by default, we associate **shmem_transport_ofi_external_heap_mrfd** with **shmem_transport_ofi_target_cntrfd** but overwrite that association in the case that the provider associates memory regions with endpoints (rather than domains)?

What does it mean for a provider to associate a memory region with an endpoint rather than a domain?

Reply:
Or I guess is it that a memory region can be associated with more than just a single completion counter or endpoint?

Reply:
You can bind endpoint or counter to ensure that any update to the memory region triggers an event on the bound object. Line number 659 is to bind the registered memory region with the target counter to ensure that any update on the memory region using `FI_REMOTE_WRITE` will generate an event to update the target counter. Line number 668 ensures binding with the endpoint when `FI_MR_ENDPOINT` is enabled. This binding is required for providers that requires this flag. From the manpages, 
```
Counter bindings are needed by providers that support the generation of completions based on fabric operations. Endpoint bindings are needed if the provider associates memory regions with endpoints (see FI_MR_ENDPOINT).
```
I am not sure about your last question... if a MR can be associated with multiple counters and/or endpoints. Need to check.